sql/src/test/java/io/crate/executor/transport/distributed/DistributingConsumerTest.java

```diff
@@ -0,0 +1,111 @@
+import org.junit.Test;
+import org.mockito.invocation.InvocationOnMock;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.*;
+
+public class DistributingConsumerTest {
+
+    private ESLogger logger = Loggers.getLogger(DistributingConsumer.class);
+
+    @Test
+    public void testSendUsingDistributingConsumerAndReceiveWithPageDownstreamContext() throws Exception {
+        Streamer<?>[] streamers = { DataTypes.INTEGER.streamer() };
+        CollectingBatchConsumer collectingConsumer = new CollectingBatchConsumer();
+        TransportDistributedResultAction distributedResultAction = mock(TransportDistributedResultAction.class);
+        PageDownstreamContext pageDownstreamContext = new PageDownstreamContext(
+            logger,
+            "n1",
+            1,
+            "dummy",
+            collectingConsumer,
+            failure -> {
+            },
+            PassThroughPagingIterator.oneShot(),
+            streamers,
+            new RamAccountingContext("dummy", new NoopCircuitBreaker("dummy")),
+            1
+        );
+        doAnswer((InvocationOnMock invocationOnMock) -> {
+            Object[] args = invocationOnMock.getArguments();
+            DistributedResultRequest resultRequest = (DistributedResultRequest) args[1];
+            ActionListener<DistributedResultResponse> listener = (ActionListener<DistributedResultResponse>) args[2];
+            resultRequest.streamers(streamers);
+            pageDownstreamContext.setBucket(0, resultRequest.rows(), resultRequest.isLast(), needMore -> {
+                listener.onResponse(new DistributedResultResponse(needMore));
+            });
+            return null;
+        }).when(distributedResultAction).pushResult(anyString(), any(), any());
+        DistributingConsumer distributingConsumer = new DistributingConsumer(
+            logger,
+            UUID.randomUUID(),
+            new ModuloBucketBuilder(streamers, 1, 0),
+            1,
+            (byte) 0,
+            0,
+            Collections.singletonList("n1"),
+            distributedResultAction,
+            streamers,
+            2, // pageSize
+            new CompletableFuture<>()
+        );
+
+        distributingConsumer.accept(TestingBatchIterators.range(0, 5), null);
+
+        List<Object[]> result = collectingConsumer.getResult();
+        assertThat(TestingHelpers.printedTable(new CollectionBucket(result)),
+            is("0\n" +
+               "1\n" +
+               "2\n" +
+               "3\n" +
+               "4\n"));
+
+        // pageSize=2 and 5 rows causes 3x pushResult
+        verify(distributedResultAction, times(3)).pushResult(anyString(), any(), any());
+    }
```

Comment:
There are some tests in DistributingDownstreamTest which no longer make sense (pause handling), but some which probably will need to be ported. Do you think it's okay to do so in the PR/commit that will remove the existing DistributingDownstream?